underscore.js

```diff
@@ -1115,6 +1115,15 @@
     if (a === b) return a !== 0 || 1 / a === 1 / b;
     // A strict comparison is necessary because `null == undefined`.
     if (a == null || b == null) return a === b;
+    // `NaN`s are equivalent, but non-reflexive.
+    if (a !== a) return b !== b;
+    // Exhaust primitive checks
+    var type = typeof a;
+    if (type !== 'function' && type !== 'object' && typeof b !== 'object') return false;
```

Comment:
Which? I think it actually adds an early exit (`(primitive, function)`).

I could be more verbose:
```js
    // Exhaust primitive checks
    var typea = typeof a;
    var typeb = typeof b;
    var eitherObjects = typea === 'object' || typeb === 'object';
    var eitherFunctions = typea === 'function' || typeb === 'function';
    var bothFunctions = typea === 'function' && typeb === 'function';
    if (eitherObjects && eitherFunctions) return false;
    if (!eitherObjects && !bothFunctions) return false;
    return deepEq(a, b, aStack, bStack);
```

Which would add all possible early exits:

| type a | type b | equals |
| -------- |  -------- |  -------- | 
| primitive | primitive | primitive checks |
| primitive | function | false |
| primitive  | object |  deepEq |
| function | primitive | **false** |
| function | function | deepEq |
| function | object | **false** |
| object | primitive | deepEq |
| object | function | false |
| object | object | deepEq |